File: PerfTester-Xamarin/PerfTester-Xamarin/MainTabBar/MainTabBarController.cs
﻿using System;
using System.IO;
using Foundation;
using PerfTesterXamarin.Screens.Home;
using PerfTesterXamarin.Screens.Info;
using PerfTesterXamarin.Screens.TestDetails;
using UIKit;

namespace PerfTesterXamarin.Screens.MainTabBar
{
    public partial class MainTabBarController : UITabBarController
    {
        public static UIColor MainColor => new UIColor(red: 0.5f, green: 0.1f, blue: 0.7f, alpha: 1f);

        string getCompileDateString()
        {
            DateTime creationDate = File.GetCreationTime("Info.plist");
            DateTime reference = TimeZone.CurrentTimeZone.ToLocalTime(new DateTime(2001, 1, 1, 0, 0, 0));
            NSDate infoDate = NSDate.FromTimeIntervalSinceReferenceDate(
                (creationDate - reference).TotalSeconds);
            NSDateFormatter dF = new NSDateFormatter();
            dF.DateFormat = "HH:mm:ss | dd.MM.yyyy";
            return dF.StringFor(infoDate);
        }

        public MainTabBarController() : base("MainTabBarController", null)
        {
        }

        public override void ViewDidLoad()
        {
            base.ViewDidLoad();
            SetupTabBar();
            SetupControllers();
        }

        private void SetupTabBar()
        {
            base.TabBar.Translucent = false;
            base.TabBar.TintColor = UIColor.White;
            base.TabBar.BarTintColor = MainColor;

            UILabel BuildDateLabel = new UILabel();
            BuildDateLabel.Text = String.Format("Compiled: {0}", getCompileDateString());
            BuildDateLabel.TextColor = UIColor.White;
            BuildDateLabel.TextAlignment = UITextAlignment.Right;
            BuildDateLabel.Font = UIFont.BoldSystemFontOfSize(10);
            TabBar.AddSubview(BuildDateLabel);
            BuildDateLabel.TranslatesAutoresizingMaskIntoConstraints = false;
            BuildDateLabel.BottomAnchor.ConstraintEqualTo(TabBar.SafeAreaLayoutGuide.BottomAnchor).Active = true;
            BuildDateLabel.LeadingAnchor.ConstraintEqualTo(TabBar.LeadingAnchor).Active = true;
            BuildDateLabel.TrailingAnchor.ConstraintEqualTo(TabBar.TrailingAnchor).Active = true;

            UILabel VersionLabel = new UILabel();
            String Version = "RELEASE";
#if DEBUG
            Version = "DEBUG";
#endif
            VersionLabel.Text = String.Format("Version: {0}", Version);
            VersionLabel.TextColor = UIColor.White;
            VersionLabel.TextAlignment = UITextAlignment.Left;
            VersionLabel.Font = UIFont.BoldSystemFontOfSize(10);
            TabBar.AddSubview(VersionLabel);
            VersionLabel.TranslatesAutoresizingMaskIntoConstraints = false;
            VersionLabel.BottomAnchor.ConstraintEqualTo(TabBar.SafeAreaLayoutGuide.BottomAnchor).Active = true;
            VersionLabel.LeadingAnchor.ConstraintEqualTo(TabBar.LeadingAnchor).Active = true;
            VersionLabel.TrailingAnchor.ConstraintEqualTo(TabBar.TrailingAnchor).Active = true;
        }

        private void SetupControllers()
        {
            //var homeCtrl = new HomeController();
            //var homeNav = new UINavigationController(homeCtrl);
            //homeNav.TabBarItem.SelectedImage = UIImage.FromBundle("HomeActive");
            //homeNav.TabBarItem.Image = UIImage.FromBundle("HomeDefault");
            //homeNav.TabBarItem.Title = "";
            //homeNav.TabBarItem.ImageInsets = new UIEdgeInsets(6, 0, -6, 0);
            //homeNav.NavigationBar.Translucent = false;
            //homeNav.NavigationBar.BarTintColor = MainColor;
            //homeNav.NavigationBar.TintColor = UIColor.White;
            //homeNav.NavigationBar.TitleTextAttributes = new UIStringAttributes() { ForegroundColor = UIColor.White };

            var testListCtrl = new TestListController();
            var testListNav = new UINavigationController(testListCtrl);
            testListNav.TabBarItem.SelectedImage = UIImage.FromBundle("TestActive");
            testListNav.TabBarItem.Image = UIImage.FromBundle("TestDefault");
            testListNav.TabBarItem.Title = "";
            testListNav.TabBarItem.ImageInsets = new UIEdgeInsets(6, 0, -6, 0);
            testListNav.NavigationBar.Translucent = false;
            testListNav.NavigationBar.BarTintColor = MainColor;
            testListNav.NavigationBar.TintColor = UIColor.White;
            testListNav.NavigationBar.TitleTextAttributes = new UIStringAttributes() { ForegroundColor = UIColor.White };

            //var infoCtrl = new InfoController();
            //var infoNav = new UINavigationController(infoCtrl);
            //infoNav.TabBarItem.SelectedImage = UIImage.FromBundle("UserActive");
            //infoNav.TabBarItem.Image = UIImage.FromBundle("UserDefault");
            //infoNav.TabBarItem.Title = "";
            //infoNav.TabBarItem.ImageInsets = new UIEdgeInsets(6, 0, -6, 0);
            //infoNav.NavigationBar.Translucent = false;
            //infoNav.NavigationBar.BarTintColor = MainColor;
            //infoNav.NavigationBar.TintColor = UIColor.White;
            //infoNav.NavigationBar.TitleTextAttributes = new UIStringAttributes() { ForegroundColor = UIColor.White };

            var navControllers = new UINavigationController[] { testListNav };
            base.ViewControllers = navControllers;

            //base.SelectedIndex = 1;
        }
    }
}

File: PerfTester-Xcode/PerfTester-Xcode/MainTabBar/MainTabBarController.swift
//
//  MainTabBarController.swift
//  PerfTester-Xcode
//
//  Created by Patryk Średziński on 08.06.2018.
//  Copyright © 2018 private.perftesterxcode. All rights reserved.
//

import UIKit

class MainTabBarController: UITabBarController {

    static var mainColor = UIColor(red: 0.1, green: 0.6, blue: 0.8, alpha: 1.0)
    
    var compileDateString: String {
        let bundleName = Bundle.main.infoDictionary!["CFBundleName"] as? String ?? "Info.plist"
        if let infoPath = Bundle.main.path(forResource: bundleName, ofType: nil),
            let infoAttr = try? FileManager.default.attributesOfItem(atPath: infoPath),
            let infoDate = infoAttr[FileAttributeKey.creationDate] as? Date {
            let dF = DateFormatter()
            dF.dateFormat = "HH:mm:ss | dd.MM.yyyy"
            return dF.string(from: infoDate)
        }
        return "Unknown"
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupTabBar()
        setupControllers()
    }

    private func setupTabBar() {
        tabBar.isTranslucent = false
        tabBar.tintColor = UIColor.white
        tabBar.barTintColor = MainTabBarController.mainColor
        
        let buildDateLabel = UILabel()
        buildDateLabel.text = "Compiled: \(compileDateString)"
        buildDateLabel.textColor = .white
        buildDateLabel.textAlignment = .right
        buildDateLabel.font = UIFont.boldSystemFont(ofSize: 10)
        tabBar.addSubview(buildDateLabel)
        buildDateLabel.translatesAutoresizingMaskIntoConstraints = false
        buildDateLabel.bottomAnchor.constraint(equalTo: tabBar.safeAreaLayoutGuide.bottomAnchor).isActive = true
        buildDateLabel.leadingAnchor.constraint(equalTo: tabBar.leadingAnchor).isActive = true
        buildDateLabel.trailingAnchor.constraint(equalTo: tabBar.trailingAnchor).isActive = true
    
        let versionLabel = UILabel()
        var version = "RELEASE"
        #if DEBUG
        version = "DEBUG"
        #endif
        versionLabel.text = "Version: \(version)"
        versionLabel.textColor = .white
        versionLabel.textAlignment = .left
        versionLabel.font = UIFont.boldSystemFont(ofSize: 10)
        tabBar.addSubview(versionLabel)
        versionLabel.translatesAutoresizingMaskIntoConstraints = false
        versionLabel.bottomAnchor.constraint(equalTo: tabBar.safeAreaLayoutGuide.bottomAnchor).isActive = true
        versionLabel.leadingAnchor.constraint(equalTo: tabBar.leadingAnchor).isActive = true
        versionLabel.trailingAnchor.constraint(equalTo: tabBar.trailingAnchor).isActive = true
    }
    
    private func setupControllers() {
//        let homeCtrl = HomeController()
//        let homeNav = UINavigationController(rootViewController: homeCtrl)
//        homeNav.tabBarItem.selectedImage = UIImage(named: "HomeActive")
//        homeNav.tabBarItem.image = UIImage(named: "HomeDefault")
//        homeNav.tabBarItem.title = ""
//        homeNav.tabBarItem.imageInsets = UIEdgeInsets.init(top: 6, left: 0, bottom: -6, right: 0)
//        homeNav.navigationBar.isTranslucent = false
//        homeNav.navigationBar.barTintColor = MainTabBarController.mainColor
//        homeNav.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor:UIColor.white]

        let testListCtrl = TestListController()
        let testListNav = UINavigationController(rootViewController: testListCtrl)
        testListNav.tabBarItem.selectedImage = UIImage(named: "TestActive")
        testListNav.tabBarItem.image = UIImage(named: "TestDefault")
        testListNav.tabBarItem.title = ""
        testListNav.tabBarItem.imageInsets = UIEdgeInsets.init(top: 6, left: 0, bottom: -6, right: 0)
        testListNav.navigationBar.isTranslucent = false
        testListNav.navigationBar.barTintColor = MainTabBarController.mainColor
        testListNav.navigationBar.tintColor = UIColor.white
        testListNav.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor:UIColor.white]

//        let infoCtrl = InfoController()
//        let infoNav = UINavigationController(rootViewController: infoCtrl)
//        infoNav.tabBarItem.selectedImage = UIImage(named: "UserActive")
//        infoNav.tabBarItem.image = UIImage(named: "UserDefault")
//        infoNav.tabBarItem.title = ""
//        infoNav.tabBarItem.imageInsets = UIEdgeInsets.init(top: 6, left: 0, bottom: -6, right: 0)
//        infoNav.navigationBar.isTranslucent = false
//        infoNav.navigationBar.barTintColor = MainTabBarController.mainColor
//        infoNav.navigationBar.tintColor = UIColor.white
//        infoNav.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor:UIColor.white]

        let navControllers = [testListNav]
        viewControllers = navControllers
        
//        selectedIndex = 1
    }

}
